jerry-debugger/src/lib/debugger-client.ts

```diff
@@ -0,0 +1,404 @@
+// limitations under the License.
+
+import assert from 'assert';
+import WebSocket from 'ws';
+
+import { Breakpoint, ParsedFunction } from './breakpoint';
+import { PROTOCOL } from './jrs-protocol-constants';
+import { FunctionMap, ParsedSource } from './parsed-source';
+import { cesu8ToString } from './string-utils';
+
+export interface JerryDebuggerOptions {
+  host?: string;
+  port?: number;
+  verbose?: boolean;
+}
+
+export const DEFAULT_DEBUGGER_HOST = 'localhost';
+export const DEFAULT_DEBUGGER_PORT = 5001;
+
+// expected JerryScript debugger protocol version
+const JERRY_DEBUGGER_VERSION = 1;
+
+export class JerryDebuggerClient {
+  readonly host: string;
+  readonly port: number;
+  readonly verbose: boolean;
+  private socket?: WebSocket;
+
+  // debugger configuration
+  private maxMessageSize: number = 0;
+  private cpointerSize: number = 0;
+  private littleEndian: boolean = true;
+  private version: number = 0;
+  private parseObj?: ParsedSource;
+  private outputResult?: Uint8Array;
+  private exceptionData?: Uint8Array;
+  private lastBreakpointHit?: object;
+  private backtraceFrame: number = 0;
+  functions: FunctionMap = {};
+
+  constructor(options: JerryDebuggerOptions) {
+    this.host = options.host || DEFAULT_DEBUGGER_HOST;
+    this.port = options.port || DEFAULT_DEBUGGER_PORT;
+    this.verbose = options.verbose || false;
+  }
+
+  connect() {
+    if (this.socket) {
+      this.disconnect();
+    }
+    this.socket = new WebSocket(`ws://${this.host}:${this.port}/jerry-debugger`);
+    this.socket.binaryType = 'arraybuffer';
+    this.socket.on('message', this.onMessage.bind(this));
+
+    return new Promise((resolve, reject) => {
+      if (!this.socket) {
+        reject();
+        return;
+      }
+
+      this.socket.on('open', () => {
+        resolve();
+      });
+
+      this.socket.on('error', (err) => {
+        reject(err);
+      });
+    });
+  }
+
+  disconnect() {
+    if (this.socket) {
+      this.socket.terminate();
+      this.socket = undefined;
+    }
+  }
+
+  abortConnection(message: string) {
+    this.disconnect();
+    console.log('Abort connection: ' + message);
+    throw new Error(message);
+  }
+
+  getFormatSize(format: string) {
+    let length = 0;
+    for (let i = 0; i < format.length; i++) {
+      switch (format[i]) {
+        case 'B':
+          length++;
+          break;
+
+        case 'C':
+          length += this.cpointerSize;
+          break;
+
+        case 'I':
+          length += 4;
+          break;
+
+        default:
+          throw new Error('unsupported message format');
+      }
+    }
+    return length;
+  }
+
+  decodeMessage(format: string, message: Uint8Array, offset = 0) {
+    // Format: B=byte I=int32 C=cpointer
+    // Returns an array of decoded numbers
+
+    const result = [];
+    let value;
+
+    if (offset + this.getFormatSize(format) > message.byteLength) {
+      this.abortConnection('received message too short');
```

Comment:
The `jerry-client-ws.py` python implementation (see `get_message`) suggests to me that it's possible to get partial messages. In other words, the parser should be able to gather and "reassemble" partial messages.